diff --git a/computed_breadcrumbs.info.yml b/computed_breadcrumbs.info.yml
index 2eea647..95e4ff4 100644
--- a/computed_breadcrumbs.info.yml
+++ b/computed_breadcrumbs.info.yml
@@ -1,5 +1,5 @@
 name: Computed Breadcrumbs
-description: Provides breadcrumbs as computed fields in nodes.
+description: Provides breadcrumbs as computed fields for content entities.
 type: module
 core_version_requirement: ^9 || ^10
 dependencies:
diff --git a/computed_breadcrumbs.module b/computed_breadcrumbs.module
index d5c2459..bfae295 100644
--- a/computed_breadcrumbs.module
+++ b/computed_breadcrumbs.module
@@ -1,5 +1,6 @@
 <?php
 
+use Drupal\Core\Entity\ContentEntityTypeInterface;
 use Drupal\Core\Entity\EntityTypeInterface;
 use Drupal\Core\Field\BaseFieldDefinition;
 use Drupal\Core\Field\FieldStorageDefinitionInterface;
@@ -9,7 +10,7 @@ use Drupal\Core\Field\FieldStorageDefinitionInterface;
  */
 function computed_breadcrumbs_entity_base_field_info(EntityTypeInterface $entity_type) {
   $fields = [];
-  if ($entity_type->id() === 'node') {
+  if ($entity_type instanceof ContentEntityTypeInterface) {
     $fields['breadcrumbs'] = BaseFieldDefinition::create('computed_breadcrumbs')
       ->setName('breadcrumbs')
       ->setLabel(t('Breadcrumbs'))
@@ -19,6 +20,7 @@ function computed_breadcrumbs_entity_base_field_info(EntityTypeInterface $entity
       ->setDisplayConfigurable('view', TRUE)
       ->setDisplayOptions('view', [
         'label' => 'hidden',
+        'region' => 'hidden',
         'weight' => -5,
       ]);
   }
diff --git a/src/Field/ComputedBreadcrumbsItemList.php b/src/Field/ComputedBreadcrumbsItemList.php
index 8761505..a432f2a 100644
--- a/src/Field/ComputedBreadcrumbsItemList.php
+++ b/src/Field/ComputedBreadcrumbsItemList.php
@@ -23,7 +23,7 @@ class ComputedBreadcrumbsItemList extends FieldItemList {
     $requestStack = \Drupal::service('request_stack');
 
     $parent = $this->getEntity();
-    if ($parent->isNew()) {
+    if ($parent->isNew() || !$parent->hasLinkTemplate('canonical')) {
       return;
     }
     $url = $parent->toUrl();
@@ -64,9 +64,13 @@ class ComputedBreadcrumbsItemList extends FieldItemList {
           if (empty($uri)) {
             $uri = 'internal:#';
           }
+          $title = $link->getText();
+          if (is_array($title)) {
+            $title = \Drupal::service('renderer')->render($title);
+          }
           $items[] = [
             'uri' => $uri,
-            'title' => $link->getText()
+            'title' => $title
           ];
         }
       });
