diff --git a/src/Resource/MenuItemsResource.php b/src/Resource/MenuItemsResource.php
index 045cc8d..65427ae 100644
--- a/src/Resource/MenuItemsResource.php
+++ b/src/Resource/MenuItemsResource.php
@@ -30,7 +30,7 @@ use Symfony\Component\Routing\Route;
  *
  * @internal
  */
-final class MenuItemsResource extends ResourceBase implements ContainerInjectionInterface {
+class MenuItemsResource extends ResourceBase implements ContainerInjectionInterface {
 
   /**
    * A list of menu items.
@@ -100,7 +100,7 @@ final class MenuItemsResource extends ResourceBase implements ContainerInjection
    * {@inheritDoc}
    */
   public static function create(ContainerInterface $container) {
-    return new self(
+    return new static(
       $container->get('menu.link_tree'),
       $container->get('entity_type.manager'),
       $container->get('entity_field.manager'),
