diff --git a/src/Controller/S3CorsUploadAjaxController.php b/src/Controller/S3CorsUploadAjaxController.php
index cd838be..59f18b2 100644
--- a/src/Controller/S3CorsUploadAjaxController.php
+++ b/src/Controller/S3CorsUploadAjaxController.php
@@ -82,7 +82,7 @@ class S3CorsUploadAjaxController extends ControllerBase {
 
     // Retrieve the file name and build the URI.
     // Destination does not contain a prefix as it is applied by the fly system.
-    $uri = \Drupal::service('file_system')->createFilename($post['filename'], $post['destination']);
+    $uri = \Drupal::service('file_system')->getDestinationFilename($post['destination']. $post['filename'], FileSystemInterface::EXISTS_RENAME);
     // Apply the prefix to the URI and use it as a key in the POST request.
     $post['key'] = $adapter->applyPathPrefix(StreamWrapperManager::getTarget($uri));
 
