diff --git a/flysystem_s3.js b/flysystem_s3.js
index ae93275..e1d4840 100644
--- a/flysystem_s3.js
+++ b/flysystem_s3.js
@@ -40,6 +40,21 @@
       // Store uploaded files fid values.
       const uploadedFileFid = [];
 
+      // Disable "other" file fields, and also submit buttons.
+      // @See Drupal.file.disableFields().
+      var $enabledFields = [];
+      if ($fileElement.closest('div.js-form-managed-file').length > 0) {
+        $enabledFields = $fileElement
+          .closest('div.js-form-managed-file')
+          .find('input.js-form-file');
+      }
+      const $fieldsToTemporarilyDisable = $(
+        'div.js-form-managed-file input.js-form-file, input[type="submit"]',
+      )
+        .not($enabledFields)
+        .not(':disabled');
+      $fieldsToTemporarilyDisable.prop('disabled', true);
+
       Object.keys(filelist).forEach(function(file) {
         const file_obj = filelist[file];
 
@@ -60,6 +75,7 @@
             );
             // Trigger the submit button to let normal AJAX process the upload.
             Drupal.file.triggerUploadButton(event);
+            $fieldsToTemporarilyDisable.prop('disabled', false);
           })
           .done(function(signedFormData) {
             Drupal.flysystemS3.setCorsUploadProgress(
@@ -78,6 +94,7 @@
                 );
                 // Trigger the submit button to let normal AJAX process the upload.
                 Drupal.file.triggerUploadButton(event);
+                $fieldsToTemporarilyDisable.prop('disabled', false);
               })
               .done(function() {
                 // Set progress bar to 100% in case the upload was so fast.
@@ -99,6 +116,7 @@
                     );
                     // Trigger the submit button to let normal AJAX process the upload.
                     Drupal.file.triggerUploadButton(event);
+                    $fieldsToTemporarilyDisable.prop('disabled', false);
                   })
                   .done(function(saveFileData) {
                     // Add the fid for this file to array.
@@ -122,6 +140,7 @@
                       // Trigger the submit button to let normal AJAX process the upload.
                       Drupal.file.triggerUploadButton(event);
                     }
+                    $fieldsToTemporarilyDisable.prop('disabled', false);
                   });
               });
           });
