diff --git a/src/Resource/MenuItemsResource.php b/src/Resource/MenuItemsResource.php
index d1ff99c..7fe83fe 100644
--- a/src/Resource/MenuItemsResource.php
+++ b/src/Resource/MenuItemsResource.php
@@ -20,7 +20,7 @@ use Symfony\Component\Routing\Route;
  *
  * @internal
  */
-final class MenuItemsResource extends ResourceBase {
+class MenuItemsResource extends ResourceBase {
 
   /**
    * A list of menu items.
