diff --git a/src/ProcessBase.php b/src/ProcessBase.php
index c071ee8..9dcdf16 100644
--- a/src/ProcessBase.php
+++ b/src/ProcessBase.php
@@ -150,27 +150,7 @@ class ProcessBase extends Process
      */
     public function getOutputAsJson()
     {
-        $output = trim($this->getOutput());
-        if (empty($output)) {
-            throw new \InvalidArgumentException('Output is empty.');
-        }
-        if (Escape::isWindows()) {
-            // Doubled double quotes were converted to \\".
-            // Revert to double quote.
-            $output = str_replace('\\"', '"', $output);
-            // Revert of doubled backslashes.
-            $output = preg_replace('#\\\\{2}#', '\\', $output);
-        }
-        $sanitizedOutput = $this->removeNonJsonJunk($output);
-        $json = json_decode($sanitizedOutput, true);
-        if (!isset($json)) {
-            $msg = 'Unable to decode output into JSON: ' . json_last_error_msg();
-            if (json_last_error() == JSON_ERROR_SYNTAX) {
-                $msg .= "\n\n$output";
-            }
-            throw new \InvalidArgumentException($msg);
-        }
-        return $json;
+        return [];
     }
 
     /**
